src/homework-12.1/index.ts

```diff
@@ -0,0 +1,146 @@
+enum GridFilterTypeEnum {
+  ExactMatch,
+  Range,
+  ValueSearch,
+}
+
+type GridFilterValue<T> = {
+  type: GridFilterTypeEnum;
+  filter: Extract<T, string | number>;
+  filterTo?: Extract<T, string | number>;
+};
+
+type GridFilterSetValues<T> = {
+  values: T[];
+};
+
+interface Movie {
+  title: string;
+  releaseYear: number;
+  rating: number;
+  awards: string[];
+}
+
+interface Category {
+  name: string;
+  movies: Movie[];
+}
+
+interface MovieFilter {
+  titleFilter?: GridFilterValue<string>;
+  releaseYearFilter?: GridFilterValue<number>;
+  ratingFilter?: GridFilterValue<number>;
+  awardsFilter?: GridFilterSetValues<string>;
+}
+
+interface CategoryFilter {
+  nameFilter?: GridFilterValue<string>;
+  moviesFilter?: MovieFilter;
+}
+
+abstract class BaseList<T, F> {
+  protected items: T[];
+  protected currentFilter: F;
+
+  constructor(items: T[]) {
+    this.items = items;
+    this.currentFilter = {} as F;
+  }
+
+  abstract applySearchValue(query: string): T[];
+  abstract applyFiltersValue(filter: F): T[];
+}
+
+class MovieList extends BaseList<Movie, MovieFilter> {
+  applySearchValue(title: string) {
+    return this.items.filter(movie => movie.title.toLowerCase().includes(title.toLowerCase()));
+  }
+
+  applyFiltersValue(filter: MovieFilter) {
```

Comment:
Please, do not forget to explicitly add type annotation for the method return type.